dataset/include/scipp/dataset/bins_view.h

```diff
@@ -0,0 +1,57 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// Copyright (c) 2020 Scipp contributors (https://github.com/scipp)
+/// @file
+/// @author Simon Heybrock
+#pragma once
+
+#include "scipp/dataset/dataset.h"
+
+namespace scipp::dataset {
+
+namespace bins_view_detail {
+template <class T, class View> class BinsCommon {
+public:
+  BinsCommon(const View &var) : m_var(var) {}
+  auto indices() const { return std::get<0>(get()); }
+  auto dim() const { return std::get<1>(get()); }
+  auto buffer() const { return std::get<2>(get()); }
+
+protected:
+  auto make(const View &view) const {
+    return make_non_owning_bins(this->indices(), this->dim(), view);
+  }
+
+private:
+  auto get() const { return m_var.template constituents<bucket<T>>(); }
+  View m_var;
+};
+
+template <class T, class View> class BinsCoords : public BinsCommon<T, View> {
+public:
+  BinsCoords(const BinsCommon<T, View> base) : BinsCommon<T, View>(base) {}
+  auto operator[](const Dim dim) const {
+    return this->make(this->buffer().coords()[dim]);
+  }
+};
+
+template <class T, class View> class Bins : public BinsCommon<T, View> {
+public:
+  using BinsCommon<T, View>::BinsCommon;
+  auto data() const { return this->make(this->buffer().data()); }
+  auto coords() const { return BinsCoords<T, View>(*this); }
+};
+} // namespace bins_view_detail
+
+/// Return helper for accessing bin data and coords as non-owning views
+///
+/// Usage:
+/// auto data = bins_view<DataArray>(var).data();
+/// auto coord = bins_view<DataArray>(var).coords()[dim];
+///
+/// The returned objects are variables referencing data in `var`. They do not
+/// own or share ownership of any data.
+template <class T, class View> auto bins_view(const View &var) {
+  return bins_view_detail::Bins<T, View>(var);
```

Comment:
Looks good, though not tested independently as far as I can see.

Reply:
Yes, not sure where we are going with this. I am considering also making it available on the Python side, e.g., as `binned.bins.coords['x']`, but not quite sure yet. Will add tests once I am using it in more places, i.e., when know know that this is the right approach.